### PR TITLE
Proposal: use isolatedModules:true in tsconfig.json to support Babel compilation

### PR DESCRIPTION
**Pull request by SimenB**
### Checklist:

- [ ] If this PR is a new feature, please reference an issue where a consensus about the design was reached (not necessary for small changes)
- [ ] Make sure all of the significant new logic is covered by tests

`isolatedModules` emulates more closely Babel's model, where each file can be transpiled in isolation without type information. Main motivation for changing in `@apollo/client` is that this allows downstream consumers to use `isolatedModules` as well without disabling type checking of `node_modules`.

This includes #6317, without which doing isolated modules is pretty much impossible. So please ignore the first commit when reviewing this one.

### CODE REVIEW

File: src/cache/inmemory/fixPolyfills.ts [hunk 1]
```diff
@@ -6,3 +6,5 @@
 // to work around problems with Map and Set polyfills in older versions of
 // React Native (which should have been fixed in react-native@0.59.0):
 // https://github.com/apollographql/apollo-client/pull/5962
+
+export {}
```
Comment (SimenB) [inline src/cache/inmemory/fixPolyfills.ts:10]: `isolatedModules` requires every file to be a module

**Comment by hwillson** [comment]
Enabling `isolatedModules` isn't something we're interested in doing at this time. You might want to consider opening an issue in our [FR repo](http://github.com/apollographql/apollo-feature-requests) to continue this discussion.

**Comment by SimenB** [comment]
That decision means consumers of `@apollo/client` cannot use `isolatedModules` either, which is unfortunate.

---

I don't have the bandwidth to open a feature request with this, especially when it takes a month to get feedback on a PR, and only 4 of the last 20 issues opened in that repo has received a comment...

(That sounds whiny, which is not my intention. I can very much empathize with not having time to go through all issues of a large OSS project!)

**Comment by benjamn** [comment]
> That decision means consumers of `@apollo/client` cannot use `isolatedModules` either, which is unfortunate.

Surely you just mean they can't use `isolatedModules` _if_ they're choosing to compile and type-check `node_modules`, which is a choice that comes with obvious consequences. They certainly _can_ use `@apollo/client`, and they can use `isolatedModules` within their application. If that's not the case, then Babel is doing something very hostile to the ecosystem by pushing this approach to TypeScript compilation.

As I understand it, the `isolatedModules` mode of `tsc` only exists for compatibility with Babel's third-party implementation of TypeScript compilation, which has a number of known limitations due to its assumption of single-file compilation (no `const enum` inlining, for example, which is a feature we use). Because Babel is popular, the TypeScript team have been forced to introduce features like `isolatedModules` that accommodate the limitations of Babel's fork of their compiler. The Apollo Client team are big believers in TypeScript, and by extension we are skeptical of other projects that would reimplement TypeScript without aiming for full compatibility.

> I don't have the bandwidth to open a feature request with this, especially when it takes a month to get feedback on a PR, and only 4 of the last 20 issues opened in that repo has received a comment...

If you're going to use arguments about your time or bandwidth, you absolutely could have opened an issue to propose this change before going ahead and putting in the work. Sorry you wasted your time.

**Comment by darkbasic** [comment]
> If that's not the case, then Babel is doing something very hostile to the ecosystem

@benjamn Babel is already damaging the Typescript ecosystem, see what happened with namespaces. IMHO burn it with fire.

**Comment by SimenB** [comment]
@benjamn sorry, missed your response (went on holiday and got some much needed rest 😀).

I don't feel like I wasted my time on _this_ PR - I went into it with eyes open that this PR could be closed without merging. I meant specifically the feature request repo. And I can see that even though I tried not to be snarky, it crept in regardless. Apologies!

**Pull request closed without merging**